sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ExtRefService.java

```diff
@@ -196,43 +191,170 @@ public static List<TExtRef> filterDuplicatedExtRefs(List<TExtRef> tExtRefs) {
+        Boolean isExtRefDescMatchDigitalChannel = tChannel.getChannelType().equals(TChannelType.DIGITAL)
+                && extRef.getDesc().startsWith("DYN_LDEPF_DIGITAL CHANNEL " + tChannel.getChannelNum()+"_1_BOOLEEN")
+                && extRef.getDesc().endsWith("_" + tChannel.getDAName() + "_1");
+        return extRef.isSetDesc() && (isExtRefDescMatchAnalogChannel || isExtRefDescMatchDigitalChannel)
+                && extRef.isSetPLN() && Utils.lnClassEquals(extRef.getPLN(), tChannel.getLNClass())
+                && extRef.isSetPDO() && extRef.getPDO().equals(tChannel.getDOName());
+    }
+
+    /**
+     * Verify whether the IED satisfies the EPF channel for the private element `TCompasICDHeader`
+     * @param iedAdapter IEDAdapter
+     * @param channel TChannel
+     * @return true if the TCompasICDHeader matches the EPF channel
+     */
+    private static boolean isIcdHeaderMatchLDEPFChannel(IEDAdapter iedAdapter, TChannel channel) {
+        return iedAdapter.getCompasICDHeader()
+                .map(compasICDHeader -> compasICDHeader.getIEDType().value().equals(channel.getIEDType())
+                        && compasICDHeader.getIEDredundancy().value().equals(channel.getIEDRedundancy().value())
+                        && compasICDHeader.getIEDSystemVersioninstance().toString().equals(channel.getIEDSystemVersionInstance()))
+                .orElse(false);
+    }
+
+    /**
+     * Provides Active LDevice according to EPF channel's inst attribute
+     * @param iedAdapter IEDAdapter
+     * @param channel TChannel
+     * @return LDeviceAdapter object that matches the EPF channel
+     */
+    private static Optional<LDeviceAdapter> getActiveSourceLDeviceByLDEPFChannel(IEDAdapter iedAdapter, TChannel channel) {
+        return iedAdapter.findLDeviceAdapterByLdInst(channel.getLDInst())
+                .filter(lDeviceAdapter -> lDeviceAdapter.getLDeviceStatus()
+                        .map(status -> status.equals(LdeviceStatus.ON.getValue()))
+                        .orElse(false))
+                .stream().findFirst();
+    }
+
+    /**
+     * Provides Active LN Object that satisfies the EPF channel attributes (lnClass, lnInst, prefix)
+     * @param lDeviceAdapter LDeviceAdapter
+     * @param channel TChannel
+     * @return AbstractLNAdapter object that matches the EPF channel
+     */
+    private static Optional<AbstractLNAdapter<?>> getActiveLNodeSourceByLDEPFChannel(LDeviceAdapter lDeviceAdapter, TChannel channel) {
+        return lDeviceAdapter.getLNAdaptersIncludingLN0().stream()
+                .filter(lnAdapter -> lnAdapter.getLNClass().equals(channel.getLNClass())
+                        && lnAdapter.getLNInst().equals(channel.getLNInst())
+                        && Utils.equalsOrBothBlank(channel.getLNPrefix(), lnAdapter.getPrefix()))
+                .findFirst()
+                .filter(lnAdapter -> lnAdapter.getDaiModStValValue()
```

Comment:
The status we checked before was for LDevice(inst=LDEPF), this time we check the status for the source LDevice.

Reply:
Here the `lDeviceAdapter `wich is used as parameter is known to be activated because checked just before; 
So we don't need to check again the `DaiModStValValue`

Reply:
You're right. the status of the source LDevice already checked, here as the name of the method mention, we verify the status of the source LN (It might not be LN0). 

I update the name of the method from `getActiveLNodeSourceByLDEPFChannel` to `getActiveLNSourceByLDEPFChannel`

Reply:
We will see next time